File: LP/practical_3_any_of_following/Kruskals_Minimal_Spanning_Tree/krushkals_mst.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

struct Edge {
    int src, dest, weight;
};

bool compare(Edge a, Edge b) {
    return a.weight < b.weight;
}

int find(vector<int>& parent, int i) {
    if (parent[i] != i)
        parent[i] = find(parent, parent[i]);
    return parent[i];
}

void unionSet(vector<int>& parent, int x, int y) {
    int xroot = find(parent, x);
    int yroot = find(parent, y);
    parent[xroot] = yroot;
}

void kruskalMST(vector<Edge>& edges, int n) {
    sort(edges.begin(), edges.end(), compare);

    vector<int> parent(n);
    for (int i = 0; i < n; i++)
        parent[i] = i;

    cout << "Edge \tWeight\n";
    for (auto& edge : edges) {
        int x = find(parent, edge.src);
        int y = find(parent, edge.dest);

        if (x != y) {
            cout << edge.src << " - " << edge.dest << "\t" << edge.weight << endl;
            unionSet(parent, x, y);
        }
    }
}

int main() {
    int n = 4; // vertices
    vector<Edge> edges = {
        {0, 1, 10},
        {0, 2, 6},
        {0, 3, 5},
        {1, 3, 15},
        {2, 3, 4}
    };

    kruskalMST(edges, n);

    return 0;
}

File: LP/practical_3_any_of_following/Kruskals_Minimal_Spanning_Tree/krushkals_MST.py
class Edge:
    def __init__(self, src, dest, weight):
        self.src = src
        self.dest = dest
        self.weight = weight

def find(parent, i):
    if parent[i] != i:
        parent[i] = find(parent, parent[i])
    return parent[i]

def union(parent, x, y):
    xroot = find(parent, x)
    yroot = find(parent, y)
    parent[xroot] = yroot

def kruskal(edges, n):
    edges.sort(key=lambda edge: edge.weight)
    parent = [i for i in range(n)]

    print("Edge \tWeight")
    for edge in edges:
        x = find(parent, edge.src)
        y = find(parent, edge.dest)

        if x != y:
            print(f"{edge.src} - {edge.dest} \t{edge.weight}")
            union(parent, x, y)

edges = [
    Edge(0, 1, 10),
    Edge(0, 2, 6),
    Edge(0, 3, 5),
    Edge(1, 3, 15),
    Edge(2, 3, 4)
]
n = 4
kruskal(edges, n)

